### ISSUE TITLE
[Flang] Incorrect diagnostic on procedure pointer dummy argument

### ISSUE THREAD

**Issue by DanielCChen** [llvm/llvm-project#126950]
Consider the following code:
```
module m
    contains

    function genF1 (i)
        procedure (real), pointer :: genF1
        integer :: i
    end function
end module

program fmisc043a
use m
    call test1(genF1(1))

    contains
    subroutine test1 (p)
        procedure(real), pointer :: p
    end subroutine
end
```

Flang is currently issue an error as
```
error: Semantic errors in t3.f
./t3.f:12:16: error: Actual argument associated with procedure pointer dummy argument 'p=' must be a pointer unless INTENT(IN)
      call test1(genF1(1))
                 ^^^^^^^^

```

The standard explicitly allows such a usage as [24-007: 15.5.2.10]
```
If a dummy argument is a procedure pointer, the corresponding actual argument shall be a procedure pointer, a
reference to a function that returns a procedure pointer, a reference to the intrinsic function NULL, or a valid
target for the dummy pointer in a pointer assignment statement.
```

**Comment by eugeneepshteyn** [comment]
Yes, seems to be a bug. The very next sentence of the standard's quote above says "If the actual argument is not a pointer, the dummy argument shall have INTENT (IN)", which may explain where `INTENT(IN)` in the error message came from, but it doesn't apply here.

**Comment by klausler** [comment]
F'2023 15.5.2.10 p5: "If a dummy argument is a procedure pointer, the corresponding actual argument shall be a procedure pointer, a reference to a function that returns a procedure pointer, a reference to the intrinsic function NULL, or a valid
target for the dummy pointer in a pointer assignment statement. *If the actual argument is not a pointer, the
dummy argument shall have INTENT (IN)*; if the actual argument is not a dummy argument it becomes pointer
associated with the actual argument, otherwise it becomes pointer associated with the ultimate argument of the
actual argument."

The text that I emphasized in the above quote is ambiguous.  Does a reference to a function qualify as a pointer in this context?  I suspect that it does, and that the requirement of `INTENT(IN)` applies only to actual arguments that are valid targets, but I think that it could also be read as applying to anything that is not a proper procedure pointer.

I'll change this to apply only to the last case (targets), but I am not confident that this is clearly correct.

**Comment by DanielCChen** [comment]
> F'2023 15.5.2.10 p5: "If a dummy argument is a procedure pointer, the corresponding actual argument shall be a procedure pointer, a reference to a function that returns a procedure pointer, a reference to the intrinsic function NULL, or a valid target for the dummy pointer in a pointer assignment statement. _If the actual argument is not a pointer, the dummy argument shall have INTENT (IN)_; if the actual argument is not a dummy argument it becomes pointer associated with the actual argument, otherwise it becomes pointer associated with the ultimate argument of the actual argument."
> 
> The text that I emphasized in the above quote is ambiguous. Does a reference to a function qualify as a pointer in this context? I suspect that it does, and that the requirement of `INTENT(IN)` applies only to actual arguments that are valid targets, but I think that it could also be read as applying to anything that is not a proper procedure pointer.
> 
> I'll change this to apply only to the last case (targets), but I am not confident that this is clearly correct.

Right, I see your point. 
The standard does say the following for `INTENT(OUT)` and `INTENT(INOUT)`
```
Any actual argument that corresponds to such a dummy pointer shall
be a pointer variable or a procedure pointer that is not the result of a function reference.
```

So I think it should be the same for all non-definable actual argument, the check should be on the presence of those two intents rather than missing `INTENT(IN)`.
